rs/rust_canisters/messaging_test/tests/messaging_test.rs

```diff
@@ -0,0 +1,167 @@
+    prop_assert_eq!(blob_size_bytes as u32, blob_size_bytes_);
+    prop_assert_eq!(payload_size_bytes, payload_size_bytes_);
+}
+
+/// Tests traffic between canisters works as intended without triggering any traps and that
+/// the instructions are adhered to faithfully.
+#[test]
+fn smoke_test() {
+    let (env1, env2) = two_subnets_simple();
+    let wasm = Project::cargo_bin_maybe_from_env("messaging-test-canister", &[]).bytes();
+
+    let canister1 = env1
+        .install_canister_with_cycles(wasm.clone(), Vec::new(), None, Cycles::new(u128::MAX / 2))
+        .expect("Installing messaging-test-canister failed");
+    let canister2 = env1
+        .install_canister_with_cycles(wasm.clone(), Vec::new(), None, Cycles::new(u128::MAX / 2))
+        .expect("Installing messaging-test-canister failed");
+    let canister3 = env2
+        .install_canister_with_cycles(wasm.clone(), Vec::new(), None, Cycles::new(u128::MAX / 2))
+        .expect("Installing messaging-test-canister failed");
+
+    // A call to be sent to `canister1` as an ingress that then calls `canister3`
+    // as a XNet inter canister call; that then makes a call to self.
+    let (receiver, payload) = to_encoded_ingress(Call {
+        receiver: canister1.into(),
+        call_bytes: 456,
+        reply_bytes: 789,
+        cycles: 1_000_000_000_000,
+        timeout_secs: None,
+        downstream_calls: vec![Call {
+            receiver: canister3.into(),
+            call_bytes: 654,
+            reply_bytes: 987,
+            cycles: 1_000_000_000,
+            timeout_secs: Some(10),
+            downstream_calls: vec![Call {
+                receiver: canister3.into(),
+                call_bytes: 123_456,
+                reply_bytes: 654_321,
+                cycles: 0,
+                timeout_secs: None,
+                downstream_calls: vec![],
+            }],
+        }],
+    });
+    let msg_id1 = env1
+        .submit_ingress_as(
+            PrincipalId::new_anonymous(),
+            receiver,
+            "handle_call",
+            payload,
+        )
+        .unwrap();
+
+    // A call to be sent to `canister2` as an ingress that then calls `canister1`
+    // on the same subnet.
+    let (receiver, payload) = to_encoded_ingress(Call {
+        receiver: canister2.into(),
+        call_bytes: 312,
+        reply_bytes: 546,
+        cycles: 500_000_000,
+        timeout_secs: Some(20),
+        downstream_calls: vec![Call {
+            receiver: canister1.into(),
+            call_bytes: 385_212,
+            reply_bytes: 2,
+            cycles: 0,
+            timeout_secs: None,
+            downstream_calls: vec![],
+        }],
+    });
+    let msg_id2 = env1
+        .submit_ingress_as(
+            PrincipalId::new_anonymous(),
+            receiver,
+            "handle_call",
+            payload,
+        )
+        .unwrap();
+
+    // Executing 20 rounds should plenty for these two calls to conclude.
```

Comment:
I suppose there is no better way to decide that we can check the success conditions than simply progressing for enough rounds?

Reply:
Agreed that it doesn't really make a difference. One could even argue that the way it is done here is easier to grasp...